### PR TITLE
feat: supportedDatasources

### LINKED ISSUE #5651: Enforce manager -> datasource links for better documentation
I think users would benefit if each manager's documentation page stated which `datasource` values it supported, e.g. `docker`, `github-tags`, etc. Ideally this would be auto-generated from the code like we do for `fileMatch` (see https://docs.renovatebot.com/modules/manager/cocoapods/ for an example).

I can think of a few ways to do this, although each has a downside:

1. Scan each manager's files for references to `datasource/*` imports and then use that to build the list. If any manager uses literal strings though, it won't catch them. Also it's a little challenging because we don't mandate file names within a manager.

2. Mandate that each `manager` export a list of datasources it supports, and then have `datasource/index` enforce that no returned deps have a datasource that mismatches. The problem here is (a) extra work (not too much..) and (b) we'd need to make sure that managers don't test themselves directly and instead test via the index function.

I guess I'd lean towards the first approach, because it should be correctly nearly all of the time but also there's not really a huge problem if documentation is temporarily wrong until someone spots it.

### LINKED ISSUE #13617: Declare supported datasources in each manager
### What would you like Renovate to be able to do?

I would like for Renovate documentation to include an accurate list of datasources supported by each manager.

### If you have any ideas on how this should be implemented, please tell us here.

The manager interface could be expanded to include a `datasources` field (string list). The challenge is how to keep it up to date, because manual updating can lead to errors.

We could possibly try to do this using linting, but I can't think of an elegant way.

An alternative would be to log a debug message whenever a manager returns an "undeclared" datasource type, and then I monitor for these in the app to detect such human errors quickly and fix them.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds `supportedDatasources` field to managers, and uses it to generate documentation.

## Context:

- Closes #13617
- Closes #5651

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/ansible/index.ts [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import { ProgrammingLanguage } from '../../constants';
+import * as datasourceDocker from '../../datasource/docker';
 import extractPackageFile from './extract';
```
Comment (viceice) [inline lib/manager/ansible/index.ts:3]:
```suggestion
export extractPackageFile from './extract';
```

File: lib/manager/ansible/index.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
 
 const language = ProgrammingLanguage.Docker;
```
Comment (viceice) [inline lib/manager/ansible/index.ts:5]:
```suggestion
export const language = ProgrammingLanguage.Docker;
```

File: lib/manager/ansible/index.ts [hunk 3]
```diff
@@ -1,9 +1,12 @@
...
 
-export { extractPackageFile, language };
+const supportedDatasources = [datasourceDocker.id];
```
Comment (viceice) [inline lib/manager/ansible/index.ts:7]:
```suggestion
export const supportedDatasources = [datasourceDocker.id];
```

File: lib/manager/ansible/index.ts [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
+
+export { extractPackageFile, language, supportedDatasources };
```
Comment (viceice) [inline lib/manager/ansible/index.ts:9]:
```suggestion
```

looks simpler

File: lib/manager/github-actions/index.ts [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import * as githubTagsDatasource from '../../datasource/github-tags';
```
Comment (viceice) [inline lib/manager/github-actions/index.ts:1]:
```suggestion
import * as githubTagsDatasource from '../../datasource/github-tags';
import * as dockerVersioning from '../../versioning/docker';
```

File: lib/manager/github-actions/index.ts [hunk 6]
```diff
@@ -8,3 +9,5 @@ export const defaultConfig = {
     '(^|\\/)action\\.ya?ml$',
   ],
 };
+
+export const supportedDatasources = [githubTagsDatasource.id];
```
Comment (viceice) [inline lib/manager/github-actions/index.ts:13]: I think we also support docker images?
```suggestion
export const supportedDatasources = [githubTagsDatasource.id, dockerVersioning.id];
```

https://github.com/renovatebot/renovate/blob/ab018191fac6961b5f9d0a73261ac8fba0b4566f/lib/manager/github-actions/extract.ts#L25-L33

File: lib/manager/jsonnet-bundler/extract.ts [hunk 7]
```diff
@@ -48,6 +48,7 @@ function extractDependency(dependency: Dependency): PackageDependency | null {
   const match = gitUrl.exec(dependency.source.git.remote);
 
   return {
+    datasource: GitRefsDatasource.id,
```
Comment (viceice) [inline lib/manager/jsonnet-bundler/extract.ts:51]: Missing import? 🙃

File: lib/manager/sbt/index.ts [hunk 8]
```diff
@@ -1,8 +1,17 @@
...
+import * as datasourceSbtPlugin from '../../datasource/sbt-plugin';
 import * as ivyVersioning from '../../versioning/ivy';
 
 export { extractPackageFile } from './extract';
 export { bumpPackageVersion } from './update';
 
+export const supportedDatasources = [
+  datasourceMaven.id,
+  datasourceSbtPackage,
+  datasourceSbtPlugin,
```
Comment (viceice) [inline lib/manager/sbt/index.ts:12]:
```suggestion
  datasourceSbtPackage.id,
  datasourceSbtPlugin.id,
```

Missing `id` ?

File: lib/manager/travis/index.ts [hunk 9]
```diff
@@ -1,10 +1,13 @@
 import { ProgrammingLanguage } from '../../constants';
+import * as datasourceGithubTags from '../../datasource/github-tags';
 import * as nodeVersioning from '../../versioning/node';
 
 export { extractPackageFile } from './extract';
 
 export const language = ProgrammingLanguage.NodeJS;
 
+export const supportedDatasources = [datasourceGithubTags.id, 'foo'];
```
Comment (viceice) [inline lib/manager/travis/index.ts:9]: `foo` ?
```suggestion
export const supportedDatasources = [datasourceGithubTags.id];
```

File: tools/docs/manager.ts [hunk 10]
```diff
@@ -56,6 +56,11 @@ sidebar_label: ${displayName}
         }
       }
       md += `For details on how to extend a manager's \`fileMatch\` value, please follow [this link](/modules/manager/#file-matching).\n\n`;
+      md += '## Supported datasources\n\n';
+      const escapedDatasources = (supportedDatasources || [])
+        .map((datasource) => `\`${datasource}\``)
+        .join(', ');
+      md += `This manager supports extracting the following datasources: ${escapedDatasources}.\n`;
```
Comment (viceice) [inline tools/docs/manager.ts:63]:
```suggestion
      md += `This manager supports extracting the following datasources: ${escapedDatasources}.\n\n`;
```

needs one more newline

We should later try to link those values

File: lib/manager/jsonnet-bundler/extract.ts (hunk 7 above)
Comment (rarkins) [inline lib/manager/jsonnet-bundler/extract.ts:51]:
```suggestion
```

File: lib/manager/travis/index.ts (hunk 9 above)
Comment (rarkins) [reply lib/manager/travis/index.ts:9]: Hmm this was me trying to test my test

File: lib/manager/regex/index.ts [hunk 11]
```diff
@@ -11,6 +11,8 @@ export const defaultConfig = {
   pinDigests: false,
 };
 
+export const supportedDatasources = ['ANY'];
```
Comment (viceice) [inline lib/manager/regex/index.ts:14]: Shouldn't we return null or empty array and add an explicit doc? 🤔

File: lib/manager/terraform/index.ts [hunk 12]
```diff
@@ -1,8 +1,19 @@
...
 import * as hashicorpVersioning from '../../versioning/hashicorp';
 
 export { updateArtifacts } from './lockfile';
 export { extractPackageFile } from './extract';
 
+export const supportedDatasources = [
+  BitBucketTagsDatasource.id,
+  GitTagsDatasource.id,
+  datasourceGithubTags.id,
+  TerraformModuleDatasource.id,
```
Comment (viceice) [inline lib/manager/terraform/index.ts:14]: Did we miss `terraform-provider`?

File: lib/manager/terragrunt/index.ts [hunk 13]
```diff
@@ -1,7 +1,16 @@
...
+import * as datasourceGithubTags from '../../datasource/github-tags';
+import { TerraformModuleDatasource } from '../../datasource/terraform-module';
 import * as hashicorpVersioning from '../../versioning/hashicorp';
 
 export { extractPackageFile } from './extract';
 
+export const supportedDatasources = [
+  GitTagsDatasource.id,
+  datasourceGithubTags.id,
+  TerraformModuleDatasource.id,
```
Comment (viceice) [inline lib/manager/terragrunt/index.ts:11]: Do we use the provider here too?

Comment (rarkins) [reply lib/manager/terragrunt/index.ts:11]: Don't think so

**Review by viceice — APPROVED** [review]

**Comment by JamieMagee** [comment]
Only 1 _super_ nitpick: you can export these as `const` assertions[^1]. For example:

```ts
export const supportedDatasources = [datasourceDocker.id]
```

becomes

```ts
export const supportedDatasources = [datasourceDocker.id] as const
```

and the type goes from

```ts
string[]
```

to

```ts
readonly ["docker"]
```

[^1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions

**Review by JamieMagee — APPROVED** [review]

**Comment by viceice** [comment]
@JamieMagee good idea, can be done for more exports. We should track this in an issue.

**Pull request merged**